### PR TITLE
Post 1.0 version policy and release process

### PR DESCRIPTION
**Pull request by verult**
Initial proposal for version policy and release process post-1.0.

The policy requires that `cirq-core` strictly follows Semantic Versioning and disallow backward-incompatible changes across minor version bumps, but leaves room for each vendor directory to make a separate choice on whether or not to maintain backward compatibility. This allows `cirq-core` to be stable while giving vendors the agility to move fast if needed.

@MichaelBroughton

### CODE REVIEW

**Comment by verult** [comment]
This is ready for review; leaving it as a draft only to prevent it from merging until wider approval.

**Comment by verult** [comment]
cc'ing vendor directory owners - would love to hear your feedback on the policy.

cirq-ionq: @dabacon @ColemanCollins @nakardo @gmauricio @Cynocracy
cirq-aqt: @ma5x @pschindler @alfrisch
cirq-pasqal: @HGSilveri
cirq-rigetti: @erichulburd @kalzoo @dbanty

Thanks in advance!

**Comment by Cynocracy** [comment]
+1 from me!

Will there be any automated checking of semantic version on the API / serializable object level?

One point that I don't fully understand, if cirq-core controls the release schedule and is bound by semver, what does the release process look like when a vendor wants to bump the major/minor independently (say, earlier than the schedule)? Is that possible?

**Comment by verult** [comment]
@Cynocracy thanks for taking a look!

> Will there be any automated checking of semantic version backcompat at the API / serializable object level?
Not at the moment, but the hope is to get to this soon after Cirq 1.0. Tracking issue: https://github.com/quantumlib/Cirq/issues/5745

> One point that I don't fully understand, if cirq-core controls the release schedule and is bound by semver, what does the release process look like when a vendor wants to bump the major/minor independently (say, earlier than the schedule)? Is that possible?
With this policy it's unfortunately not possible. The main complexity is around the version compatibility matrix if vendor versions were to differ from `cirq-core`. The policy allows this to be changed if we do figure out the compatibility story in the future: "Packages may have different version numbers in the future, at which point this policy will be updated." Let me know if the policy wording should be clarified in any way.

That being said, there may be rare occasions when an urgent patch is necessary for any of the packages, and releasing from the main development branch will be unstable. I'm drafting a one-off process we could potentially follow, but in the mean time open to suggestions!

**Review by Cynocracy — APPROVED** [review]
One concern I would have as a user/developer is formally defining "backward compatibility". All previously working invocations must work? No deleted methods? It's easy to get side tracked in discussions about what we mean by compatibility without formalizing it.

LGTM, I appreciate the flexibility this grants vendors!

File: release.md
Comment (verult) [reply release.md:53]: I'm not sure about this one - IMO a released backward-breaking change which is discovered later should be considered a bug and be fixed in a subsequent release. Although an enforcement mechanism such as CI is a lot more accurate than enforcement through maintainer reviews, any system could have bugs that let some breaking changes through, and I think the stronger "are not allowed" should be used when either of the two mechanisms are in place.

WDYT?

**Comment by verult** [comment]
> One concern I would have as a user/developer is formally defining "backward compatibility". All previously working invocations must work? No deleted methods? It's easy to get side tracked in discussions about what we mean by compatibility without formalizing it.

Yeah totally agree this is very important and these discussions could easily come up with future code changes. I've also tracked this question in https://github.com/quantumlib/Cirq/issues/5745, and this needs to be addressed soon so as to not block PR progress after Cirq 1.0.

**Comment by verult** [comment]
Since urgent patches will happen rarely, the process won't be included in this initial version of the policy and. The effort to document a process is tracked in https://github.com/quantumlib/Cirq/issues/5745

File: release.md
Comment (HGSilveri) [inline release.md:42]: So, if I understand correctly, the version of the vendor packages will still track the version of `cirq-core`, right?

File: release.md
Comment (HGSilveri) [inline release.md:71]: I'm ok with this approach, though it is unclear how things would unfold should a vendor package require an urgent patch because it seems that `cirq-core` dictates the release schedule. I agree it should be a rare event, but I would still like to see it documented in the "one-off process" you mentioned above.

**Review by HGSilveri — COMMENTED** [review]
+1 from me too, I only have a couple remarks/questions

File: release.md
Comment (verult) [reply release.md:42]: Thanks for your review! Yes that's right.

File: release.md
Comment (verult) [reply release.md:71]: Agree that this process needs to be documented. I have a proposal (in short, making a minor version release branch when a patch is necessary) but it still needs to be discussed with a wider audience. Because this is rare, the plan is to defer the discussion to unblock releasing the version policy.

**Comment by verult** [comment]
Hi everyone, we are aiming to include the policy in the Cirq 1.0 release - if you'd like to chime in on the policy, it would be greatly appreciated if you could leave a comment by EOD today Pacific time. Thanks to everyone who has reviewed so far!

**Pull request merged**
